adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -725,45 +1207,151 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+      // it is more efficient this way because we avoid a shuffle operation
+      // and we have to go through all the data anyway
+      .mapPartitions(iter => {
+        iter.flatMap(_._2).toList.sortBy(_._2).toIterator
+      }, preservesPartitioning = true)
+
+    // here we get the new partition map
+    // for now, we can't use the left RDD's partition map because
+    // the bounds don't always include the values we copartitioned with.
+    // in other words, joins allow data that starts before a given
+    // ReferenceRegion to combine given they overlap.
+    // this means that we cannot use the bounds of the left RDD because
+    // they may not neccessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getPartitionMapBoundsFromRdd(iter.map(_.swap))
+    }, preservesPartitioning = true)
+
+    replaceRdd(finalPartitionedRDD.keys, Some(newPartitionMap))
+  }
+
+  /**
+    * Gets the partition bounds from a ReferenceRegion keyed Iterator
+    *
+    * @param iter The data on a given partition. ReferenceRegion keyed
+    * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+    */
+  private def getPartitionMapBoundsFromRdd(iter: Iterator[(ReferenceRegion, T)]): Iterator[(ReferenceRegion, ReferenceRegion)] = {
+    if (iter.isEmpty) Iterator()
+    else {
+      val head = iter.next._1
+      if (iter.hasNext) {
+        var tail = iter.next._1
+        while (iter.hasNext) tail = iter.next._1
+        Iterator((head, tail))
+        //only one ReferenceRegion on this partition
+      } else Iterator((head, head))
+    }
+  }
+
+  /**
+   * This method is the first method called when there is some need for the
+   * data to be sorted. We are assuming no previous knowledge about data
+   * distribution or skew, and as such we simply partition based on range.
+   * This can present some issues when the data is extremely skewed, but
+   * we attempt to do our best in solving this problem.
+   *
+   * The output from this method is a SortedGenomicRDD, which simply
+   * contains the knowledge that this was sorted and overrides many methods
+   * that would be improved by sorting the data first.
+   *
+   * @param partitions The number of partitions to split the data into
+   * @return A SortedGenomicRDDMixIn that contains the sorted and partitioned
+   *         RDD
+   */
+  def repartitionAndSortByGenomicCoordinate(partitions: Int = rdd.partitions.length)(implicit c: ClassTag[T]): U = {
+    // here we let spark do the heavy lifting
+    val partitionedRDD = flattenRddByRegions().sortBy(f => f._1, ascending = true, partitions)
+    val newPartitionMapRdd = partitionedRDD.mapPartitions(iter => getPartitionMapBoundsFromRdd(iter), preservesPartitioning = true)
+    replaceRdd(partitionedRDD.values, Some(newPartitionMapRdd))
+  }
+
+  private[rdd] class GenomicPositionRangePartitioner[V](partitions: Int, elements: Int = 0) extends Partitioner {
```

Comment:
`GenomicPosition` is used in [`GenomicPartitioners.scala`](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicPartitioners.scala#L39).